### ISSUE TITLE
[RFC] Want a method like Java's 'Optional#orElseGet()' in Data.Optional

### ISSUE THREAD

**Issue by tyru** [vim-jp/vital.vim#576]
I want to call a function **when optional is none.**
But I don't want to write `:if`.

NOTE: Java's `Optional#orElseGet(Supplier<T> f)`: `f` is a function (-like instance) and it is invoked when optional instance is empty.
The method returns a return value of `f`.
https://docs.oracle.com/javase/jp/8/docs/api/java/util/Optional.html#orElseGet-java.util.function.Supplier-

### Expected

```vim
" s:f() returns optional.
call s:O.get_or(s:f(), {-> execute('echo "failed"', '')})
```

### Current

```vim
let o = s:f()
if s:O.empty(o)
  echo "failed"'
endif
```

# What is problem?

I want to confirm some points below:

* Is there already a function like this?
* Should we extend the 2nd argument of `s:O.get_or()` like above expected code?

**Comment by ujihisa** [comment]
> * Is there already a function like this?

Unfortunately there isn't. You can combine with map() and get_or() to achieve this, but it's not straightforward. e.g.

```vim
let s:O = g:V.import('Data.Optional')

function! s:default() abort
  echo 'calling default()'
  return 123
endfunction

" plain old way
function! s:aaa(x) abort
  if s:O.empty(a:x)
    return s:default()
  else
    return s:O.get_unsafe(a:x) " s:O.get() also works but you don't have to
  endif
endfunction

" method chained way
function! s:bbb(x) abort
  return s:O.get_or(s:O.map(a:x, {x -> {-> x}}), function('s:default'))()
endfunction

echo '## aaa'
echo s:aaa(s:O.some(1))
echo s:aaa(s:O.none())
echo ''
echo '## bbb'
echo s:bbb(s:O.some(1))
echo s:bbb(s:O.none())

" results
## aaa
1
calling default()
123

## bbb
1
calling default()
123
```

> * Should we extend the 2nd argument of s:O.get_or() like above expected code?

Should be provided as another function, such as `get_or_lazy` or with better name.

**Comment by tyru** [comment]
@ujihisa ah, wrapping a value with function is smart way. thanks.

**Comment by ujihisa** [comment]
Yes, but there should be alternative straightforward way just by a single function call ;)

**Comment by tyru** [comment]
implemented by #610
